### ISSUE TITLE
[Bug]: Genesis poor performance during simple benchmarking test

### ISSUE THREAD

**Issue by BigRorzUni** [Genesis-Embodied-AI/Genesis#1456]
### Bug Description

Hi there, I am an intern running some benchmark tests on genesis. I tried to do similar tests to the ones in https://github.com/zhouxian/genesis-speed-benchmark/blob/main/genesis_corrected_franka_benchmark.py 

However, Genesis starts performing more poorly than I would've expected (outperformed by MJX for all batch sizes) and I am not experienced enough with it to be entirely sure as to why this is or if I have something configured incorrectly.

My test essentially runs an increasing number of steps 10000 -> 100000000 on a simple xml file. It does this for 3 different batch sizes/no. environments (2048, 4096 & 8192).

Here is the xml file:
```
<mujoco>
<worldbody>
    <light name="top" pos="0 0 1"/>
    <body name="box_and_sphere" euler="0 0 -30">
    <joint name="swing" type="hinge" axis="1 -1 0" pos="-.2 -.2 -.2"/>
    <geom name="red_box" type="box" size=".2 .2 .2" rgba="1 0 0 1"/>
    <geom name="green_sphere" pos=".2 .2 .2" size=".1" rgba="0 1 0 1"/>
    </body>
</worldbody>
</mujoco>
```

Here is my python script:
```
import genesis as gs
import numpy as np
import time

import argparse
import timing_helper

parser = argparse.ArgumentParser(description="Run Genesis Pendulum Simulation")

parser.add_argument("input_lb", type=int, help="Lower bound of input range")
parser.add_argument("input_ub", type=int, help="Upper bound of input range")
parser.add_argument("input_points", type=int, help="Number of input points")
parser.add_argument("-B", type=int, default=2048) # batch size

args = parser.parse_args()

n_envs = args.B

def simulate(scene, total_steps):
    print()
    print("Warmup")

    num_batch_steps = int(total_steps / scene.n_envs)
    print(f'Num steps per Env: {num_batch_steps}')
    
    for i in range(200): 
        scene.step()

    print("Warmup done, now testing")

    t0 = time.perf_counter()
    for _ in range(num_batch_steps):
        scene.step()
    t1 = time.perf_counter()

    t = t1-t0 
    fps_per_env = round(1000 / t, 2)
    total_fps = fps_per_env * n_envs
    print(f'per env: {fps_per_env:,.2f} FPS')
    print(f'total  : {total_fps:,.2f} FPS')'


    print(f'this took : {t1-t0} seconds')

    return t, fps_per_env, total_fps

def main():
    print(f"Lower Bound: {args.input_lb}")
    print(f"Upper Bound: {args.input_ub}")
    print(f"Input Points: {args.input_points}")
    print(f"Batch Size: {args.B}")
    print("-----------------------------")

    inputs = np.logspace(args.input_lb, args.input_ub, args.input_points)
    inputs = [int(x) for x in inputs]

    print("Setting up scene")

    gs.init(backend=gs.gpu, logging_level='warning')

    scene = gs.Scene(
        show_viewer=False,
        rigid_options=gs.options.RigidOptions(
            dt=0.01,
        ),
    )

    scene.add_entity(gs.morphs.MJCF(file="../xml/pendulum.xml"))
    
    scene.build(n_envs)

    times = []
    fps_per_env = []
    total_fps = []

    for steps in inputs:
        t, e_fps, t_fps = simulate(scene, steps)

        times.append(t)
        fps_per_env.append(e_fps)
        total_fps.append(t_fps)

    timing_helper.send_times_csv(inputs, times, f"data/Genesis/{n_envs}_speed.csv", "Speed (s)")
    timing_helper.send_times_csv(inputs, fps_per_env, f"data/Genesis/{n_envs}_env_fps.csv", "FPS")
    timing_helper.send_times_csv(inputs, total_fps, f"data/Genesis/{n_envs}_total_fps.csv", "FPS")


if __name__ == "__main__":
    main()
```
Here is my nvidia-smi output and CUDA installs in case these shine a light on the situation
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3090        Off | 00000000:01:00.0  On |                  N/A |
|  0%   41C    P8              37W / 350W |    976MiB / 24576MiB |     10%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```
```
# --------------- Install CuDa / nvcc ----------------
# Base installer
wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2404/x86_64/cuda-ubuntu2404.pin
mv cuda-ubuntu2404.pin /etc/apt/preferences.d/cuda-repository-pin-600
wget https://developer.download.nvidia.com/compute/cuda/12.4.0/local_installers/cuda-repo-ubuntu2404-12-4-local_12.4.0-550.54.14-1_amd64.deb
dpkg -i cuda-repo-ubuntu2404-12-4-local_12.4.0-550.54.14-1_amd64.deb
cp /var/cuda-repo-ubuntu2404-12-4-local/cuda-*-keyring.gpg /usr/share/keyrings/
apt-get update
apt-get -y install cuda-toolkit-12-4

# Driver installer (possibly untested)
apt-get install -y cuda-drivers

# -----------------    Install Cudnn    --------------------
# Some prerequisite
apt-get install zlib1g
# Main installation
wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2404/x86_64/cuda-keyring_1.1-1_all.deb
dpkg -i cuda-keyring_1.1-1_all.deb
apt-get update
apt-get -y install cudnn9-cuda-12
```

Here are the plotted results, with no. physics steps on the x-axis, time taken to execute those steps (s) on the y-axis and the different simulator + batch sizes plotted:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/e38ccc5b-c86b-46b5-9673-e5e009c6e046" />

### Steps to Reproduce

If possible, provide a script triggering the bug, e.g.
```python
# code snippet triggering the bug
import genesis as gs
gs.init()
scene = gs.Scene(show_viewer=False)
broken = scene.add_entity(
    gs.morphs.MJCF(file='attachement.xml'),
)
scene.build()
```
Make sure to attached any needed assets (here `attachement.xml`)!

### Expected Behavior

I would have expected genesis to outperform MJX significantly for all batch sizes

### Relevant log output

```shell

```

### Environment

- OS: [e.g. Ubuntu 24.04, Windows 11 24H2]
- GPU/CPU [e.g. A100, RTX 4090, M3pr, Intel I9-9900k, Ryzen 5900x] (N/A if no GPU/CPU)
- GPU-driver version (N/A if no GPU)
- CUDA / CUDA-toolkit version (N/A if non-Nvidia)

### Release version or Commit ID

release 0.2.1

**Comment by YilingQiao** [comment]
- where to get `timing_helper` package?
- what's your command to run this benchmark? 
- What's the script for MJX?

**Comment by YilingQiao** [comment]
And in your plot, larger batch sizes use less time? For both mjx and genesis?

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/9c87f3b3-f778-4bdf-8398-e6aa9e3595f7" />

**Comment by BigRorzUni** [comment]
> * where to get `timing_helper` package?

The 'timing_helper` package is code to send the data into a csv file, here is the relevant function:
```
def send_times_csv(inputs, times, fname, label_prefix, batch_sizes=None, input_prefix="steps"):
    """
    Writes timing data to a CSV file. If multiple timing batches are provided,
    each gets a separate column. Otherwise, it's written as a single time series.

    Parameters:
    - inputs: list of step counts
    - times: list of timing lists (or a single timing list)
    - fname: output CSV filename
    - label_prefix: used in column headers (e.g. 'GPU' or 'CPU Parallel')
    - batch_sizes: list of batch sizes if applicable, or None
    """

    data = {input_prefix: inputs}

    if batch_sizes:
        for i, batch_size in enumerate(batch_sizes):
            col_name = f"{label_prefix} - Batch size {batch_size} (s)"
            data[col_name] = times[i]
    else:
        # Normalize times to a list of lists if there is only one batch
        times = [times]
        data[f"{label_prefix}"] = times[0]

    df = pd.DataFrame(data)
    df.to_csv(fname, index=False)

    print(f"Timing data written to {fname}")
```
>   * what's your command to run this benchmark?

To run this benchmark I have a small shell script 
```
INPUT_LB=5
INPUT_UB=9
INPUT_POINTS=5

BATCH_SIZES=(2048 4096 8192)

for B in "${BATCH_SIZES[@]}"
    do
        echo "Running simulation with batch size $B"
        python test_scripts/genesis_pendulum.py $INPUT_LB $INPUT_UB $INPUT_POINTS -B "$B"
    done
```
>    * What's the script for MJX?

The MJX test script is as follows:
```
import mujoco
from mujoco import mjx

import jax
from jax import numpy as jp
import numpy as np

jax.config.update("jax_enable_x64", False)

import time
import math
import timing_helper 
import argparse

import os

parser = argparse.ArgumentParser(description="Run MJX Pendulum Simulation")

parser.add_argument("input_lb", type=int, help="Lower bound of input range")
parser.add_argument("input_ub", type=int, help="Upper bound of input range")
parser.add_argument("input_points", type=int, help="Number of input points")
parser.add_argument("-B", type=int, default=2048) # batch size

args = parser.parse_args()

###### will boost performance by 30% according to mjx documentation ######
xla_flags = os.environ.get('XLA_FLAGS', '')
xla_flags += ' --xla_gpu_triton_gemm_any=True'
os.environ['XLA_FLAGS'] = xla_flags

n_envs = args.B
jit_steps = 1

def time_model(mj_model, steps):
    print()
    print("Warmup")

    mj_data = mujoco.MjData(mj_model)

    mjx_model = mjx.put_model(mj_model)
    mjx_data = mjx.put_data(mj_model, mj_data)
    mjx_data = jax.vmap(lambda _: mjx_data)(np.arange(n_envs))

    print(f'Num total steps {steps}')
    print(f'num envs: {n_envs}')

    num_gpu_steps = math.ceil(steps / n_envs)
    print(f'Num Steps per Env: {num_gpu_steps}')

    def mjx_step_n(model, data):
        for _ in range(jit_steps):
            data = mjx.step(model, data)
        return data
    
    jit_step_n = jax.jit(jax.vmap(mjx_step_n, in_axes=(None, 0)), backend='gpu')

    curr_step = 0
    while True:
        mjx_data = jit_step_n(mjx_model, mjx_data)
        curr_step += jit_steps
        if curr_step >= 200:
            break
    print("Warmup done, now testing")

    t0 = time.perf_counter()
    curr_step = 0
    while True:
        mjx_data = jit_step_n(mjx_model, mjx_data)

        curr_step += jit_steps
        if curr_step >= num_gpu_steps:
            break

    t1 = time.perf_counter()

    t = t1-t0 
    fps_per_env = round(1000 / t, 2)
    total_fps = fps_per_env * n_envs
    print(f'per env: {fps_per_env:,.2f} FPS')
    print(f'total  : {total_fps:,.2f} FPS')
    print(f'this took : {t1-t0} seconds')

    return t, fps_per_env, total_fps

    

def main():
    print(f"Lower Bound: {args.input_lb}")
    print(f"Upper Bound: {args.input_ub}")
    print(f"Input Points: {args.input_points}")
    print(f"Batch Size: {args.B}")

    inputs = np.logspace(args.input_lb, args.input_ub, args.input_points)
    inputs = [int(x) for x in inputs]

    mj_model = mujoco.MjModel.from_xml_path("../xml/pendulum.xml")

    mj_model.opt.solver = 1 
    mj_model.opt.timestep = 0.01

    times = []
    fps_per_env = []
    total_fps = []

    for steps in inputs:
        t, e_fps, t_fps = time_model(mj_model, steps)
        
        times.append(t)
        fps_per_env.append(e_fps)
        total_fps.append(t_fps)
    
    timing_helper.send_times_csv(inputs, times, f"data/MJX/{n_envs}_speed.csv", f"MJX Time GPU - Batch size {n_envs} (s)")
    timing_helper.send_times_csv(inputs, fps_per_env, f"data/MJX/{n_envs}_env_fps.csv", f"MJX FPS GPU - Batch size {n_envs}")
    timing_helper.send_times_csv(inputs, total_fps, f"data/MJX/{n_envs}_total_fps.csv", f"MJX FPS GPU - Batch size {n_envs}")

if __name__ == "__main__":
    main()
```
>    * And in your plot, larger batch sizes use less time? For both mjx and genesis?

Sorry, I wasn't fully clear about how my test works. I run a total number of steps, say 10000, across all batch sizes. So if I had a batch size of 2048 then that would be 10000/2048 = ~5 steps per environment. Then for batch size 4096 there would instead be 10000/4096 = ~2 steps per environment. So larger batch sizes tend to use less time across all simulators I have tested

**Comment by duburcqa** [comment]
I expect the performance to be better now. Do not forget to enable `performance_mode=True` when initialising Genesis if you want to max out performance. This is especially important for "small" batch sizes (i.e. < 4096).

**Comment by duburcqa** [comment]
Closing because inactivity.
